AirCasting/Models/AirBeamCellularSessionCreator.swift

```diff
@@ -0,0 +1,95 @@
+        guard let sessionType = sessionContext.sessionType,
+              let sessionUUID = sessionContext.sessionUUID else {
+            assertionFailure("invalidCreateSessionContext \(sessionContext)")
+            completion(.failure(AirBeamSessionCreatorError.invalidCreateSessionContext(sessionContext)))
+            return
+        }
+        // Save data to app's database
+        let session = Session(uuid: sessionUUID,
+                              type: sessionType,
+                              name: sessionContext.sessionName,
+                              deviceType: sessionContext.deviceType,
+                              location: sessionContext.startingLocation,
+                              startTime: Date(),
+                              followedAt: Date(),
+                              tags: sessionContext.sessionTags)
+        
+        // if session is fixed: create an empty session on server,
+        // then send AB auth data to connect to web session and data needed to start recording
+        guard let name = session.name,
+              let startTime = session.startTime,
+              let peripheral = sessionContext.peripheral,
+              let contribute = sessionContext.contribute,
+              let isIndoor = sessionContext.isIndoor else {
+            assertionFailure("invalidCreateSessionContext \(sessionContext)")
+            completion(.failure(AirBeamSessionCreatorError.invalidCreateSessionContext(sessionContext)))
+            return
+        }
+        
+        #warning("TODO: change mocked data -->  notes")
+        let params = CreateSessionApi.SessionParams(uuid: sessionUUID,
+                                                    type: .fixed,
+                                                    title: name,
+                                                    tag_list: session.tags ?? "",
+                                                    start_time: startTime,
+                                                    end_time: startTime,
+                                                    contribute: contribute,
+                                                    is_indoor: isIndoor,
+                                                    notes: [],
+                                                    version: 0,
+                                                    streams: [:],
+                                                    latitude: sessionContext.startingLocation?.latitude,
+                                                    longitude: sessionContext.startingLocation?.longitude)
+        
+        createSessionService.createEmptyFixedWifiSession(input: .init(session: params,
+                                                                      compression: true),
+                                                         completion: { [measurementStreamStorage, userAuthenticationSession] result in
+                                                            DispatchQueue.main.async {
```

Comment:
I'm not a fan of dispatching to main thread in objects other than views. Couple of main reasons:
* If you dispatch completion to main it will suddenly get much much harder to stub and test
* It's really a UIKit requirement, why would some low-level "creator" type object know about it after all? ([single responsibility principle](https://stackify.com/solid-design-principles/))
* The more we just casually dispatch to main instead of keeping stuff on background threads the quicker we'll run into UI lagging/cluttering issues. Also energy impact will get pretty high (it's already high for our app)

Please take a look if this dispatch can be moved out of here, preferably to a View. If it cannot be done easily - a warning to do so in the future would be nice

Reply:
I investigated a bit and I think it cannot be done easily. I'm adding the warning there, and the task in Trello

Reply:
bummer, this main thread abuse is starting to come out ;) we should put it high on the list of things to do after mvp relase